components/raftstore/src/store/worker/split_controller.rs

```diff
@@ -0,0 +1,532 @@
+            cfg_tracker: config_manager.0.clone().tracker("split_hub".to_owned()),
+        }
+    }
+
+    pub fn default() -> AutoSplitController {
+        AutoSplitController::new(SplitConfigManager::default())
+    }
+
+    pub fn flush(&mut self, others: Vec<QpsStats>) -> (Vec<usize>, Vec<SplitInfo>) {
+        let mut split_infos = Vec::default();
+        let mut top = BinaryHeap::with_capacity(TOP_N as usize);
+
+        let mut region_info_mat = HashMap::default();
+        let capacity = others.len();
+        for other in others {
+            for (region_id, region_info) in other.region_infos {
+                if region_info.key_ranges.len() >= self.cfg.sample_num {
+                    let region_infos = region_info_mat
+                        .entry(region_id)
+                        .or_insert_with(|| Vec::with_capacity(capacity));
+                    region_infos.push(region_info);
+                }
+            }
+        }
+
+        for (region_id, region_infos) in region_info_mat {
+            let pre_sum = prefix_sum(region_infos.iter(), RegionInfo::get_qps);
+
+            let qps = *pre_sum.last().unwrap(); // region_infos is not empty
+            let num = self.cfg.detect_times;
+            if qps > self.cfg.qps_threshold {
+                let recorder = self
+                    .recorders
+                    .entry(region_id)
+                    .or_insert_with(|| Recorder::new(num));
+
+                recorder.update_peer(&region_infos[0].peer);
+
+                let key_ranges = sample(
+                    self.cfg.sample_num,
+                    &pre_sum,
+                    region_infos,
+                    RegionInfo::get_key_ranges_mut,
+                );
+
+                recorder.record(key_ranges);
+                if recorder.is_ready() {
+                    let key = recorder.collect(&self.cfg);
+                    if !key.is_empty() {
+                        let split_info = SplitInfo {
+                            region_id,
+                            split_key: Key::from_raw(&key).into_encoded(),
+                            peer: recorder.peer.clone(),
+                        };
+                        split_infos.push(split_info);
+                        info!("load base split region";"region_id"=>region_id);
+                    }
+                    self.recorders.remove(&region_id);
+                }
+            } else {
+                self.recorders.remove_entry(&region_id);
+            }
+            top.push(qps);
+        }
+
+        (top.into_vec(), split_infos)
+    }
+
+    pub fn clear(&mut self) {
+        let interval = Duration::from_secs(self.cfg.detect_times * 2);
+        self.recorders
+            .retain(|_, recorder| recorder.create_time.elapsed().unwrap() < interval);
+    }
+
+    pub fn refresh_cfg(&mut self) {
+        if let Some(incoming) = self.cfg_tracker.any_new() {
+            self.cfg = incoming.clone();
+        }
+    }
```

Comment:
There' no lock here, are these two functions called in the same thread with `flush`?

Reply:
Yes, the same thread